add modular pcl packages for ubuntu and debian

Aposhian opened the pull request:
Please add the following dependency to the rosdep database.

## Package name:
- libpcl-common
- libpcl-features
- libpcl-filters
- libpcl-io
- libpcl-kdtree
- libpcl-keypoints
- libpcl-ml
- libpcl-octree
- libpcl-outofcore
- libpcl-people
- libpcl-recognition
- libpcl-registration
- libpcl-sample-consensus
- libpcl-search
- libpcl-segmentation
- libpcl-stereo
- libpcl-surface
- libpcl-tracking
- libpcl-visualization

## Package Upstream Source:

https://github.com/PointCloudLibrary/pcl

## Purpose of using this:
Allow Ubuntu and Debian users to only pull in the dependencies they need, since all of PCL is quite large (including GUI dependencies).

Other platforms which don't have these modular packages will simply fallback to their catchall packages, maintaining compatibility

Distro packaging links:

## Links to Distribution Packages

- Debian: https://packages.debian.org/
  - https://packages.debian.org/stable/libpcl-apps1.11
  - https://packages.debian.org/oldstable/libpcl-apps1.9
  - https://packages.debian.org/stable/libpcl-common1.11
  - https://packages.debian.org/oldstable/libpcl-common1.9
  - https://packages.debian.org/stable/libpcl-features1.11
  - https://packages.debian.org/oldstable/libpcl-features1.9
  - https://packages.debian.org/stable/libpcl-filters1.11
  - https://packages.debian.org/oldstable/libpcl-filters1.9
  - https://packages.debian.org/stable/libpcl-io1.11
  - https://packages.debian.org/oldstable/libpcl-io1.9
  - https://packages.debian.org/stable/libpcl-kdtree1.11
  - https://packages.debian.org/oldstable/libpcl-kdtree1.9
  - https://packages.debian.org/stable/libpcl-keypoints1.11
  - https://packages.debian.org/oldstable/libpcl-keypoints1.9
  - https://packages.debian.org/stable/libpcl-ml1.11
  - https://packages.debian.org/oldstable/libpcl-ml1.9
  - https://packages.debian.org/stable/libpcl-octree1.11
  - https://packages.debian.org/oldstable/libpcl-octree1.9
  - https://packages.debian.org/stable/libpcl-outofcore1.11
  - https://packages.debian.org/oldstable/libpcl-outofcore1.9
  - https://packages.debian.org/stable/libpcl-people1.11
  - https://packages.debian.org/oldstable/libpcl-people1.9
  - https://packages.debian.org/stable/libpcl-recognition1.11
  - https://packages.debian.org/oldstable/libpcl-recognition1.9
  - https://packages.debian.org/stable/libpcl-registration1.11
  - https://packages.debian.org/oldstable/libpcl-registration1.9
  - https://packages.debian.org/stable/libpcl-sample-consensus1.11
  - https://packages.debian.org/oldstable/libpcl-sample-consensus1.9
  - https://packages.debian.org/stable/libpcl-search1.11
  - https://packages.debian.org/oldstable/libpcl-search1.9
  - https://packages.debian.org/stable/libpcl-segmentation1.11
  - https://packages.debian.org/oldstable/libpcl-segmentation1.9
  - https://packages.debian.org/stable/libpcl-stereo1.11
  - https://packages.debian.org/oldstable/libpcl-stereo1.9
  - https://packages.debian.org/stable/libpcl-surface1.11
  - https://packages.debian.org/oldstable/libpcl-surface1.9
  - https://packages.debian.org/stable/libpcl-tracking1.11
  - https://packages.debian.org/oldstable/libpcl-tracking1.9
  - https://packages.debian.org/stable/libpcl-visualization1.11
  - https://packages.debian.org/oldstable/libpcl-visualization1.9
- Ubuntu: https://packages.ubuntu.com/
   - https://packages.ubuntu.com/bionic/libpcl-apps1.8
   - https://packages.ubuntu.com/focal/libpcl-apps1.10
   - https://packages.ubuntu.com/jammy/libpcl-apps1.12
   - https://packages.ubuntu.com/bionic/libpcl-common1.8
   - https://packages.ubuntu.com/focal/libpcl-common1.10
   - https://packages.ubuntu.com/jammy/libpcl-common1.12
   - https://packages.ubuntu.com/bionic/libpcl-features1.8
   - https://packages.ubuntu.com/focal/libpcl-features1.10
   - https://packages.ubuntu.com/jammy/libpcl-features1.12
   - https://packages.ubuntu.com/bionic/libpcl-filters1.8
   - https://packages.ubuntu.com/focal/libpcl-filters1.10
   - https://packages.ubuntu.com/jammy/libpcl-filters1.12
   - https://packages.ubuntu.com/bionic/libpcl-io1.8
   - https://packages.ubuntu.com/focal/libpcl-io1.10
   - https://packages.ubuntu.com/jammy/libpcl-io1.12
   - https://packages.ubuntu.com/bionic/libpcl-kdtree1.8
   - https://packages.ubuntu.com/focal/libpcl-kdtree1.10
   - https://packages.ubuntu.com/jammy/libpcl-kdtree1.12
   - https://packages.ubuntu.com/bionic/libpcl-keypoints1.8
   - https://packages.ubuntu.com/focal/libpcl-keypoints1.10
   - https://packages.ubuntu.com/jammy/libpcl-keypoints1.12
   - https://packages.ubuntu.com/bionic/libpcl-ml1.8
   - https://packages.ubuntu.com/focal/libpcl-ml1.10
   - https://packages.ubuntu.com/jammy/libpcl-ml1.12
   - https://packages.ubuntu.com/bionic/libpcl-octree1.8
   - https://packages.ubuntu.com/focal/libpcl-octree1.10
   - https://packages.ubuntu.com/jammy/libpcl-octree1.12
   - https://packages.ubuntu.com/bionic/libpcl-outofcore1.8
   - https://packages.ubuntu.com/focal/libpcl-outofcore1.10
   - https://packages.ubuntu.com/jammy/libpcl-outofcore1.12
   - https://packages.ubuntu.com/bionic/libpcl-people1.8
   - https://packages.ubuntu.com/focal/libpcl-people1.10
   - https://packages.ubuntu.com/jammy/libpcl-people1.12
   - https://packages.ubuntu.com/bionic/libpcl-recognition1.8
   - https://packages.ubuntu.com/focal/libpcl-recognition1.10
   - https://packages.ubuntu.com/jammy/libpcl-recognition1.12
   - https://packages.ubuntu.com/bionic/libpcl-registration1.8
   - https://packages.ubuntu.com/focal/libpcl-registration1.10
   - https://packages.ubuntu.com/jammy/libpcl-registration1.12
   - https://packages.ubuntu.com/bionic/libpcl-sample-consensus1.8
   - https://packages.ubuntu.com/focal/libpcl-sample-consensus1.10
   - https://packages.ubuntu.com/jammy/libpcl-sample-consensus1.12
   - https://packages.ubuntu.com/bionic/libpcl-search1.8
   - https://packages.ubuntu.com/focal/libpcl-search1.10
   - https://packages.ubuntu.com/jammy/libpcl-search1.12
   - https://packages.ubuntu.com/bionic/libpcl-segmentation1.8
   - https://packages.ubuntu.com/focal/libpcl-segmentation1.10
   - https://packages.ubuntu.com/jammy/libpcl-segmentation1.12
   - https://packages.ubuntu.com/bionic/libpcl-stereo1.8
   - https://packages.ubuntu.com/focal/libpcl-stereo1.10
   - https://packages.ubuntu.com/jammy/libpcl-stereo1.12
   - https://packages.ubuntu.com/bionic/libpcl-surface1.8
   - https://packages.ubuntu.com/focal/libpcl-surface1.10
   - https://packages.ubuntu.com/jammy/libpcl-surface1.12
   - https://packages.ubuntu.com/bionic/libpcl-tracking1.8
   - https://packages.ubuntu.com/focal/libpcl-tracking1.10
   - https://packages.ubuntu.com/jammy/libpcl-tracking1.12
   - https://packages.ubuntu.com/bionic/libpcl-visualization1.8
   - https://packages.ubuntu.com/focal/libpcl-visualization1.10
   - https://packages.ubuntu.com/jammy/libpcl-visualization1.12
- Fedora: https://packages.fedoraproject.org/
  - https://packages.fedoraproject.org/pkgs/pcl/pcl/
- Arch: https://www.archlinux.org/packages/
  - not available (only in https://aur.archlinux.org/packages/pcl)
- Gentoo: https://packages.gentoo.org/
  - https://packages.gentoo.org/packages/sci-libs/pcl
- macOS: https://formulae.brew.sh/
  - https://formulae.brew.sh/formula/pcl
- Alpine: https://pkgs.alpinelinux.org/packages
  - https://pkgs.alpinelinux.org/package/edge/testing/x86_64/pcl-libs
- NixOS/nixpkgs: https://search.nixos.org/packages
  - https://search.nixos.org/packages?show=pcl
- openSUSE: https://software.opensuse.org/package/
  - not available (only in https://build.opensuse.org/package/show/science/pcl)

Aposhian committed:
add modular pcl packages for ubuntu and debian

Aposhian committed:
package name typo

methylDragon commented:
Nosetests are failing

Aposhian committed:
add missing arch placeholders

Aposhian committed:
fix debian alphabetical order

Aposhian committed:
fix bionic typo

Aposhian committed:
remove arch and opensuse modular pcl

wjwwood reviewed (dismissed):
lgtm, needs a second review

ivanpauno commented:
@Aposhian please provide links for each package for each distro, so they can be checked

mvieth committed:
add homebrew and alpine packages

ivanpauno commented on rosdep/base.yaml:
```diff
@@ -4626,6 +4626,321 @@ libpcl-all-dev:
   slackware: [pcl]
   ubuntu: [libpcl-dev]
+libpcl-apps:
+  debian:
```
```suggestion
  arch: [pcl]
  debian:
```

ivanpauno replied:
please apply to all the other libs this change

mvieth replied:
In my pull request, a check complained that the pcl package could not be found on arch and opensuse, see here: https://github.com/ros/rosdistro/pull/35137#issuecomment-1309476280

ivanpauno replied:
Thanks for the info.
I was looking at the preexisting `libpcl-all-dev` rule.
We should fix that one later, but no need to do that here.

ivanpauno commented on rosdep/base.yaml:
```diff
@@ -4626,6 +4626,321 @@ libpcl-all-dev:
+  macports: [libpcl]
+  nixos: [pcl]
+  openembedded: [pcl@meta-ros-common]
+  rhel: [pcl, pcl-tools]
```
same:
```suggestion
  rhel: [pcl]
```

ivanpauno commented on rosdep/base.yaml:
```diff
@@ -4626,6 +4626,321 @@ libpcl-all-dev:
+  gentoo: [sci-libs/pcl]
+  macports: [libpcl]
+  nixos: [pcl]
+  openembedded: [pcl@meta-ros-common]
```
```suggestion
  openembedded: [pcl@meta-ros-common]
  opensuse: [pcl]
```

ivanpauno replied:
See https://github.com/ros/rosdistro/pull/35289#discussion_r1025180890, resolving

ivanpauno reviewed (changes requested):
Please apply the suggestions to all the other packages

ivanpauno commented:
> May be a dup of https://github.com/ros/rosdistro/pull/35137, I will double check when reviewing again

It's not exactly a dup, but there's some duplication.
I will merge the first one that's ready, the other one will need a rebase after that.

mvieth commented on rosdep/base.yaml:
```diff
@@ -4626,6 +4626,321 @@ libpcl-all-dev:
   rhel: [pcl-devel]
   slackware: [pcl]
   ubuntu: [libpcl-dev]
+libpcl-apps:
```
You could add `alpine: [pcl-libs]` as well, and also the packages on homebrew like here: https://github.com/ros/rosdistro/pull/35137/files#diff-86e444436e8b275218b0f9fce40452b3a5a24dad5f9bc554b8ec9737134a3684

mvieth reviewed (commented):
Hi, I am the author of the other pull request (https://github.com/ros/rosdistro/pull/35137) which follows a similar idea. I am willing to close my pull request in favour of this one if you make the suggested changes

Aposhian commented:
Ok. Thanks for pointing out the sibling PR @ivanpauno. Best I can tell the differences are:
- this PR adds some additional packages
- this PR does not yet add macOS config
- some inconsistencies with how arch, alpine, and opensuse is handled

I'm happy to resolve the requested tasks, but it may take me a few days.

Happy to see there are others who don't want to pull in all of PCL at a time!

Aposhian committed:
remove pcl-tools from most libpcl rules

Aposhian committed:
add alpine pcl-libs to remaining libpcl rules

Aposhian commented:
Ok I have:
- added homebrew config from the other PR
- added `alpine: [pcl-libs]` for everything.

I think the last outstanding thing is to add package links to the PR description?

Aposhian commented:
@ivanpauno @mvieth I believe I have made all requested changes. Let me know if there was anything I missed.

mvieth commented on rosdep/osx-homebrew.yaml:
```diff
@@ -315,6 +315,34 @@ libpcl-all-dev:
   osx:
     homebrew:
       packages: [pcl]
+libpcl-common:
```
Would be good if you also add the rules for the other packages here: libpcl-apps, libpcl-keypoints, libpcl-ml, libpcl-octree, libpcl-outofcore, libpcl-people, libpcl-recognition, libpcl-registration, libpcl-sample-consensus, libpcl-search, libpcl-stereo, libpcl-tracking, libpcl-visualization

Aposhian replied:
Done

mvieth commented on rosdep/base.yaml:
```diff
@@ -4626,6 +4626,341 @@ libpcl-all-dev:
+  openembedded: [pcl@meta-ros-common]
+  rhel: [pcl]
+  slackware: [pcl]
+  ubuntu:
```
What about bionic for libpcl-segmentation, libpcl-stereo, libpcl-surface, libpcl-tracking, libpcl-visualization?

Aposhian committed:
add missing bionic rules

Aposhian committed:
add missing homebrew pcl rules

ivanpauno commented on rosdep/osx-homebrew.yaml:
```diff
@@ -222,8 +222,6 @@ libflann-dev:
 libfltk-dev:
   osx:
     homebrew:
-      packages: [fltk]
-libftdi-dev:
```
what did happen here?
seems that you should restore those lines.



CI output:

> 225:3 [key-duplicates] duplication of key "osx" in mapping

Aposhian replied:
Yes that looks like a typo. Will fix.

ivanpauno commented on rosdep/osx-homebrew.yaml:
```diff
@@ -315,6 +313,86 @@ libpcl-all-dev:
+  osx:
+    homebrew:
+      packages: [pcl]
+libpcl-keypoints:
+  osx:
+    homebrew:
+      packages: [pcl]
+libpcl-kdtree:
+  osx:
+    homebrew:
+      packages: [pcl]
```
alpha order:

```suggestion
libpcl-kdtree:
  osx:
    homebrew:
      packages: [pcl]
libpcl-keypoints:
  osx:
    homebrew:
      packages: [pcl]
```

ivanpauno reviewed (commented):
A few more changes to go, otherwise lgtm

Aposhian committed:
fix alphabetical order

Co-authored-by: Ivan Santiago Paunovic <ivanpauno@ekumenlabs.com>

Aposhian committed:
re-add missing lines

ivanpauno reviewed (approved):
LGTM!

ivanpauno commented:
There has been many changes from the first review, so I will wait for a second review to be done.
Thanks @Aposhian @mvieth for the PR!!

Aposhian commented:
It looks like we have two approvals. Can we get a merge on this and close it out?
